### PR TITLE
[TS SDK] script to fetch and publish ans contract for tests

### PR DESCRIPTION
**Pull request by 0xmaayan**
### Description

TS SDK supports ANS and we want to write tests for it. Since ANS contract is not part of aptos-framework we need to get the ANS contract, publish it to local testnet and test against it.

Adding a TS script to
- clone ans contract repo
- publish it on local testnet
- run tests against it
- delete cloned repo folder

We run this script whenever we run `pnpm test` in the TS SDK.

### CODE REVIEW

File: ecosystem/typescript/sdk/scripts/publish_ans_contract.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  console.log("---get default profile info---");
+  const data = execSync(GET_DEFAULT_PROFILE_COMMAND, {
+    cwd: __dirname + ANS_CORE_FOLDER,
+  })
+    .toString()
+    .trim();
+
+  // 4. get default profile account address
+  console.log("---get default profile account address---");
+  const storedData = JSON.parse(data).Result.default.account;
```
Comment (banool) [inline ecosystem/typescript/sdk/scripts/publish_ans_contract.ts:47]: This is an address right, could we rename this variable to be more descriptive?

File: ecosystem/typescript/sdk/package.json [hunk 2]
```diff
@@ -22,7 +22,7 @@
     "_build:browser": "tsup --platform browser --format iife --global-name aptosSDK --minify",
     "_build:node": "tsup --format cjs,esm --dts",
     "lint": "eslint \"**/*.ts\"",
-    "test": "jest",
+    "test": "ts-node ./scripts/publish_ans_contract.ts && jest",
```
Comment (banool) [inline ecosystem/typescript/sdk/package.json:25]: idk if we want to do this here right, then we publish every time even if you don't need to? Maybe this should be a separate command? Not super opinionated on this though.

**Review by banool — APPROVED** [review]

File: ecosystem/typescript/sdk/scripts/publish_ans_contract.ts (hunk 1 above)
Comment (0xmaayan) [reply ecosystem/typescript/sdk/scripts/publish_ans_contract.ts:47]: yes, my bad. copypasta

File: ecosystem/typescript/sdk/package.json (hunk 2 above)
Comment (0xmaayan) [reply ecosystem/typescript/sdk/package.json:25]: we do want to run this script every time we run tests - otherwise tests would fail.

Comment (banool) [reply ecosystem/typescript/sdk/package.json:25]: I mean if you make a local testnet, you only need to run this once right. But you can run the tests with `jest` multiple times after that. I imagine this is what you do when you're locally developing these tests?

Comment (0xmaayan) [reply ecosystem/typescript/sdk/package.json:25]: gotcha.

**Review by banool — APPROVED** [review]
Beauuuuuuuuuuuuuuuutiful

**Review by angieyth — APPROVED** [review]

File: .github/workflows/sdk-release.yaml [hunk 3]
```diff
@@ -38,6 +38,7 @@ jobs:
       APTOS_NODE_URL: http://127.0.0.1:8080/v1
       APTOS_FAUCET_URL: http://127.0.0.1:8081
       FAUCET_AUTH_TOKEN: ${{ secrets.DEVNET_TAP_AUTH_TOKEN }}
+      APTOS_INVOCATION: docker run ${{ secrets.GCP_DOCKER_ARTIFACT_REPO }}/tools:${{ inputs.GIT_SHA }} aptos"
```
Comment (banool) [inline .github/workflows/sdk-release.yaml:41]: I think you've got a rogue `"` here.

File: ecosystem/typescript/sdk/scripts/publish_ans_contract.ts [hunk 4]
```diff
@@ -2,8 +2,10 @@ const path = require("path");
 const { execSync } = require("child_process");
 
 const ANS_CORE_FOLDER = "/aptos-names-contracts/core";
-const APTOS_INIT_COMMAND = "aptos init --network local";
-const GET_DEFAULT_PROFILE_COMMAND = "aptos config show-profiles --profile default";
+const APTOS_INVOCATION = process.env.APTOS_INVOCATION || "aptos";
```
Comment (banool) [inline ecosystem/typescript/sdk/scripts/publish_ans_contract.ts:5]: A comment explaining what this is for would be great.

**Review by banool — APPROVED** [review]
Nice work this came out really nice!

File: .github/workflows/sdk-release.yaml (hunk 3 above)
Comment (0xmaayan) [reply .github/workflows/sdk-release.yaml]: haha CI just told me that!

File: ecosystem/typescript/sdk/scripts/publish_ans_contract.ts (hunk 4 above)
Comment (0xmaayan) [reply ecosystem/typescript/sdk/scripts/publish_ans_contract.ts:5]: absolutely! I need to make some clean ups before landing it, thanks!

File: .github/workflows/sdk-release.yaml (hunk 3 above)
Comment (banool) [reply .github/workflows/sdk-release.yaml]: Btw you might need the `--network host` flag too. But try it without it first and see what happens.

File: ecosystem/typescript/sdk/scripts/publish_ans_contract.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ */
+
+// on local publishing we want to use `aptos` commnads and on CI we want to use `docker`
+const APTOS_INVOCATION = process.env.APTOS_INVOCATION || "aptos";
+// environment we use when testing
+const APTOS_NODE_URL = process.env.APTOS_NODE_URL;
+const APTOS_FAUCET_URL = process.env.APTOS_FAUCET_URL;
+// ans account we use to publish the contract
+const ANS_TEST_ACCOUNT_PRIVATE_KEY = process.env.ANS_TEST_ACCOUNT_PRIVATE_KEY;
+const ANS_TEST_ACCOUNT_ADDRESS = process.env.ANS_TEST_ACCOUNT_ADDRESS;
```
Comment (banool) [inline ecosystem/typescript/sdk/scripts/publish_ans_contract.ts:22]: Now that `pnpm test` requires these env vars to be set, could we either add defaults here or add instructions to the README saying that you must set these? I think the defaults option could be simplest.

**Review by banool — APPROVED** [review]
Thanks for adding those default values to `publish_ans_contract.ts`! Just to confirm, this means that I can run `pnpm test` locally without having to worry about any of the ANS stuff now? If so, looks good!

**Comment by 0xmaayan** [comment]
> Thanks for adding those default values to `publish_ans_contract.ts`! Just to confirm, this means that I can run `pnpm test` locally without having to worry about any of the ANS stuff now? If so, looks good!

I do plan to add the `publish_ans_contract` script to run when we run `pnpm test` - 
1. The ans script is part of the full sdk tests, and I think separate it to a new and different command and expect people to remember to run it when they run pnpm test can easily become a mess. 
2. `pnpm test` should mostly run once locally before `git push` and on CI. To run tests against one change, it is recommended to run a single test file (and not the full tests)

**Comment by banool** [comment]
Sure that's all fine, I agree it should be one command, I was just asking if this adds any extra complexity from the user side, like needing to set more args or anything. If not, then this is good.

**Comment by 0xmaayan** [comment]
> Sure that's all fine, I agree it should be one command, I was just asking if this adds any extra complexity from the user side, like needing to set more args or anything. If not, then this is good.

ohh yes, no extra complexity. all env vars have default values

**Pull request merged**